lib/Dialect/Torch/Transforms/RefineTypes.cpp

```diff
@@ -981,55 +981,41 @@ ChangeResult TypeAnalyzer::visitReshapeLikeOp(
-  auto knowledge =
-      ValueKnowledge::getNotNonePessimisticValueState(op.getContext());
-  knowledge.dtype = input.dtype;
-  knowledge.hasSizes = input.hasSizes;
-  auto dim0 = op.dim0();
-  auto dim1 = op.dim1();
-  int64_t dim0Int;
-  int64_t dim1Int;
-  if (matchPattern(dim0, m_TorchConstantInt(&dim0Int)) &&
-      matchPattern(dim1, m_TorchConstantInt(&dim1Int))) {
-    knowledge.sizes = input.sizes;
-    int64_t inputRank = input.sizes.size();
-    dim0Int = toPositiveDim(dim0Int, inputRank);
-    dim1Int = toPositiveDim(dim1Int, inputRank);
-    if (isValidDim(dim0Int, inputRank) && isValidDim(dim1Int, inputRank)) {
-      std::swap(knowledge.sizes[dim0Int], knowledge.sizes[dim1Int]);
-      return getLatticeElement(op.getResult()).join(knowledge);
-    }
-  }
-
-  knowledge.sizes.resize(input.sizes.size(), kUnknownSize);
-  return getLatticeElement(op.getResult()).join(knowledge);
-}
-
-ChangeResult TypeAnalyzer::visitAtenPermuteOp(
-    AtenPermuteOp op, ArrayRef<LatticeElement<ValueKnowledge> *> operands) {
+template <typename OpTy>
+ChangeResult TypeAnalyzer::visitPermutationLikeOps(
+    OpTy op, ArrayRef<LatticeElement<ValueKnowledge> *> operands) {
   auto input = operands[0]->getValue();
   auto knowledge =
       ValueKnowledge::getNotNonePessimisticValueState(op.getContext());
   knowledge.dtype = input.dtype;
   if (!input.hasSizes)
     return getLatticeElement(op.getResult()).join(knowledge);
   knowledge.hasSizes = input.hasSizes;
+  int64_t inputRank = input.sizes.size();
   knowledge.sizes.resize(input.sizes.size(), kUnknownSize);
-  Value dims = op.dims();
-  SmallVector<int64_t> dimensions;
-  if (matchPattern(dims, m_TorchConstantIntList(dimensions))) {
-    int64_t inputRank = input.sizes.size();
-    for (unsigned i = 0; i < inputRank; i++) {
-      int64_t dim = toPositiveDim(dimensions[i], inputRank);
-      if (!isValidDim(dim, inputRank)) {
-        return getLatticeElement(op.getResult()).join(knowledge);
-      }
-      knowledge.sizes[i] = input.sizes[dim];
+  SmallVector<int64_t> dimensions(input.sizes.size());
+  if (isa<AtenTransposeIntOp>(op)) {
+    std::iota(dimensions.begin(), dimensions.end(), 0);
+    int64_t dim0Int, dim1Int;
+    auto transposeInt = cast<AtenTransposeIntOp>(op);
+    if (!matchPattern(transposeInt.dim0(), m_TorchConstantInt(&dim0Int)) ||
+        !matchPattern(transposeInt.dim1(), m_TorchConstantInt(&dim1Int)))
+      return getLatticeElement(transposeInt.getResult()).join(knowledge);
+    dimensions[toPositiveDim(dim0Int, inputRank)] = dim1Int;
+    dimensions[toPositiveDim(dim1Int, inputRank)] = dim0Int;
+  } else if (isa<AtenPermuteOp>(op)) {
+    auto permute = cast<AtenPermuteOp>(op);
+    if (!matchPattern(permute.dims(), m_TorchConstantIntList(dimensions)))
+      return getLatticeElement(permute.getResult()).join(knowledge);
+  } else {
+    return getLatticeElement(op.getResult()).join(knowledge);
+  }
+  for (unsigned i = 0; i < inputRank; i++) {
```

Comment:
I personally prefer to keep the RefineTypes separate for the 2 ops.  This loop is not very efficient for `AtenTransposeIntOp` because for `AtenTransposeIntOp` only 2 dimensions need to check for validity and they need to be checked above when changing `dimensions`. What do you think?